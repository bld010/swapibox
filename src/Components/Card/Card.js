import React from 'react';
import PropTypes from 'prop-types';
import './Card.scss'

const Card = (props) => {

  let { name, 
    species, 
    homeworld, 
    population, 
    language, 
    terrain, 
    model, 
    vehicle_class, 
    residents, 
    climate, 
    type,
    passengers,
    handleFavorite,
    favorites
   } = props



   let getFavoriteInfo = () => {

    if (type === 'people') {
      return { name, species, homeworld, population, language, type}
    }

    if (type === 'vehicles') {
      return { name, model, vehicle_class, passengers, type}
    }

    if (type === 'planets') {
      return { name, population, terrain, climate, residents, type}
    }

  }

  let peopleCard = 
    <>
    <h5>| Species: </h5> 
    <p>{species}</p>
    <h5>| Home: </h5>
    <p>{homeworld} (Pop. {population})</p>
    <h5>| Language: </h5>
    <p>{language}</p>
    </>;

  let residentsList = 
    <>
      <p className='residents'></p>
      <span>{residents && residents.length > 0 && residents.join(', ')}</span>
      {residents && residents.length === 0 && <p>None</p>}
    </>
  
  let planetCard = 
    <>
    <h5>| Terrain: </h5>
    <p>{terrain}</p>
    <h5>| Population: </h5>
    <p>{population}</p>
    <h5>| Climate: </h5>
    <p>{climate}</p>
    <h5>| Residents: </h5>
    {residentsList}
    </>;
  
  let vehicleCard = 
    <>
    <h5>| Model: </h5>
    <p>{model} </p>
    <h5>| Class: </h5>
    <p>{vehicle_class}</p>
    <h5>| Passengers: </h5>
    <p>{passengers}</p>
    </>;

  
  let getCardClass = () => {
    if (favorites.map(favorite => favorite.name).includes(name)) {
      return 'favorite'
    } else {
      return ''
    }
  }
  

  let handleFavoriteButton = (favoritedCard) => {
    handleFavorite(favoritedCard);
  }

  return(
    <article className={type + ' Card'}>
      <button className={getCardClass()} onClick={() => handleFavoriteButton(getFavoriteInfo())}></button>
        
      <h4>{name}</h4>
      
      {type === 'people' && peopleCard}
      {type === 'planets' && planetCard}
      {type === 'vehicles' && vehicleCard}
      
    </article>
  )
}

export default Card;

Card.propTypes = {
  type: PropTypes.string,
  name: PropTypes.string,
  species:  PropTypes.string,
  homeworld:  PropTypes.string,
  population:  PropTypes.string,
  language:  PropTypes.string,
  terrain:  PropTypes.string,
  model:  PropTypes.string,
  vehicle_class:  PropTypes.string,
  residents: PropTypes.array,
  climate:  PropTypes.string,
  passengers:  PropTypes.string,
  handleFavorite: PropTypes.func,
  favorites: PropTypes.array
}