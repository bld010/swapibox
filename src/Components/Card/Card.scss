$main-color: green;
$name-color: goldenrod;
$box-shadow: 0px 0px 8px 3px rgb(67, 165, 71);

.Card {
  height: 250px;
  width: 250px;
  border: 1px solid $main-color;
  color: $main-color;
  background: black;
  margin: 10px 30px 30px 10px;
  box-shadow: $box-shadow;
  padding: 20px;
  border-radius: 0 29px 0 29px;
  text-align: left;

  & h4 {
    color: $name-color;
    font-size: 22px;
    text-transform: uppercase;
    
  } 

  & p {
    font-size: 18px;
    text-transform: uppercase;
    font-weight: bold;
  }

  & button {
    margin-left: 220px;
    border-radius: 50%;
    border: 1px solid $main-color;
    height: 30px;
    width: 30px;
    background: black;
    cursor: pointer
  }
}