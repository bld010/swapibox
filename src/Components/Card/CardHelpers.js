export let getFavoriteInfo = (props) => {

  let { name, 
    species, 
    homeworld, 
    population, 
    language, 
    terrain, 
    model, 
    vehicle_class, 
    residents, 
    climate, 
    type,
    passengers,
   } = props

  if (type === 'people') {
    return { name, species, homeworld, population, language, type}
  }

  if (type === 'vehicles') {
    return { name, model, vehicle_class, passengers, type}
  }

  if (type === 'planets') {
    return { name, population, terrain, climate, residents, type}
  }

}

export let getCardClass = (favorites, name) => {
  if (favorites.map(favorite => favorite.name).includes(name)) {
    return 'favorite'
  } else {
    return ''
  }
}