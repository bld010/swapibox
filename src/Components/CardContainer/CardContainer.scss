$background-color: #000000;

.CardContainer {
  background: $background-color;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  padding: 50px;
  min-height: 500px;
  margin: 0;
  font-family: 'Courier';
}

.CardContainer.favorites {
  font-size: 30px;
  color: green;
}