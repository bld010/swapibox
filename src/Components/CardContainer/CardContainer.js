import React from 'react';
import PropTypes from 'prop-types';
import Card from '../Card/Card';
import './CardContainer.scss'

const CardContainer = ({ cards, type, handleFavorite, favorites, error }) => {
  return(
    <section className={type + " CardContainer"}>
      {cards !== null && cards.map(card => 
        <Card key={card.name} favorites={favorites} handleFavorite={handleFavorite} type={type} {...card}/>
      )}
      { error !== null && cards === null && <p>{error}</p>}
      {error === null && cards === null && <p>Loading ...</p>}

    </section>
  )
}

export default CardContainer;

CardContainer.propTypes = {
  cards: PropTypes.array,
  type: PropTypes.string,
  handleFavorite: PropTypes.func,
  favorites: PropTypes.array,
  error: PropTypes.string
}