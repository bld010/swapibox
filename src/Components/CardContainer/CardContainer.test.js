import React from 'react';
import CardContainer from './CardContainer';
import { shallow } from 'enzyme';
import { Card } from '../Card/Card';

import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

let mockCards = [
  {name: "Luke Skywalker", homeworld: "Tatooine", population: "200000", species: "Human", language: "Galactic Basic"},
  {name: "C-3PO", homeworld: "Tatooine", population: "200000", species: "Droid", language: "n/a"},
  {name: "R2-D2", homeworld: "Naboo", population: "4500000000", species: "Droid", language: "n/a"}
]

let mockHandleFavorite = jest.fn();

let mockFavorites = [
  {name: "Luke Skywalker", homeworld: "Tatooine", population: "200000", species: "Human", language: "Galactic Basic"},
]

let error = null;

describe('CardContainer', () => {
  it('should match the wrapper with data passed through', () => {
    const wrapper = shallow(<CardContainer 
      cards={mockCards}
      type='people'
      handleFavorite={mockHandleFavorite}
      favorites={mockFavorites}
      error={error} />)

      expect(wrapper).toMatchSnapshot();
    
  })
})

