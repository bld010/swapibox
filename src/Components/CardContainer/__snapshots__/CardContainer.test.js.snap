// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardContainer should match the wrapper with data passed through 1`] = `
<section
  className="people CardContainer"
>
  <Card
    favorites={
      Array [
        Object {
          "homeworld": "Tatooine",
          "language": "Galactic Basic",
          "name": "Luke Skywalker",
          "population": "200000",
          "species": "Human",
        },
      ]
    }
    handleFavorite={[MockFunction]}
    homeworld="Tatooine"
    key="Luke Skywalker"
    language="Galactic Basic"
    name="Luke Skywalker"
    population="200000"
    species="Human"
    type="people"
  />
  <Card
    favorites={
      Array [
        Object {
          "homeworld": "Tatooine",
          "language": "Galactic Basic",
          "name": "Luke Skywalker",
          "population": "200000",
          "species": "Human",
        },
      ]
    }
    handleFavorite={[MockFunction]}
    homeworld="Tatooine"
    key="C-3PO"
    language="n/a"
    name="C-3PO"
    population="200000"
    species="Droid"
    type="people"
  />
  <Card
    favorites={
      Array [
        Object {
          "homeworld": "Tatooine",
          "language": "Galactic Basic",
          "name": "Luke Skywalker",
          "population": "200000",
          "species": "Human",
        },
      ]
    }
    handleFavorite={[MockFunction]}
    homeworld="Naboo"
    key="R2-D2"
    language="n/a"
    name="R2-D2"
    population="4500000000"
    species="Droid"
    type="people"
  />
</section>
`;
